src/shogun/base/macros.h

```diff
@@ -20,8 +20,95 @@
 
 // a quick macro for making sure that an object
 // does not have a copy-ctor and operator=
-#define SG_DELETE_COPY_AND_ASSIGN(TypeName) \
-	TypeName(const TypeName&) = delete; \
+#define SG_DELETE_COPY_AND_ASSIGN(TypeName)                                    \
+	TypeName(const TypeName&) = delete;                                        \
 	void operator=(const TypeName&) = delete
 
+#ifdef _MSC_VER
+
+#define VA_NARGS(...)                                                          \
+	INTERNAL_EXPAND_ARGS_PRIVATE(INTERNAL_ARGS_AUGMENTER(__VA_ARGS__))
+#define INTERNAL_ARGS_AUGMENTER(...) unused, __VA_ARGS__
+#define INTERNAL_EXPAND(x) x
+#define INTERNAL_EXPAND_ARGS_PRIVATE(...)                                      \
+	INTERNAL_EXPAND(INTERNAL_GET_ARG_COUNT_PRIVATE(                            \
+	    __VA_ARGS__, 8, 7, 6, 5, 4, 3, 2, 1, 0))
+#define INTERNAL_GET_ARG_COUNT_PRIVATE(                                        \
+    _0_, _1_, _2_, _3_, _4_, _5_, _6_, _7_, _8_ count, ...)                    \
+	count
+
+#else
+
+#define VA_NARGS_IMPL(_1, _2, _3, _4, _5, _6, _7, _8, N, ...) N
+#define VA_NARGS(...) VA_NARGS_IMPL(__VA_ARGS__, 8, 7, 6, 5, 4, 3, 2, 1)
+
+#endif
+
+#define VARARG_IMPL2(base, count, ...) base##count(__VA_ARGS__)
+#define VARARG_IMPL(base, count, ...) VARARG_IMPL2(base, count, __VA_ARGS__)
+#define VARARG(base, ...) VARARG_IMPL(base, VA_NARGS(__VA_ARGS__), __VA_ARGS__)
```

Comment:
@karlnapf I added these macros to this header because I need them for SG_ADD_OPTIONS

Reply:
Sure thing